File: modules/avatar-view/android/src/main/java/expo/modules/avatarview/AvatarView.kt
package expo.modules.avatarview

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Typeface
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.views.ExpoView

class AvatarView(context: Context, appContext: AppContext) : ExpoView(context, appContext) {

  private var name: String = ""
  private var sizeDp: Int = 40

  private val backgroundPaint = Paint(Paint.ANTI_ALIAS_FLAG)
  private val textPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
    color = Color.WHITE
    typeface = Typeface.create(Typeface.DEFAULT, Typeface.BOLD)
    textAlign = Paint.Align.CENTER
  }

  /** Deterministic color palette — mirrors the JS Avatar component */
  private val avatarColors = intArrayOf(
    Color.parseColor("#0D9488"), // teal (accent)
    Color.parseColor("#7C3AED"), // purple
    Color.parseColor("#2563EB"), // blue
    Color.parseColor("#EF4444"), // red
    Color.parseColor("#F59E0B"), // amber
    Color.parseColor("#10B981"), // green
  )

  fun setName(newName: String) {
    name = newName
    backgroundPaint.color = colorFromName(newName)
    invalidate()
  }

  fun setSize(newSize: Int) {
    sizeDp = newSize
    invalidate()
  }

  override fun onDraw(canvas: Canvas) {
    super.onDraw(canvas)
    val w = width.toFloat()
    val h = height.toFloat()
    val radius = minOf(w, h) / 2f

    // Draw circle background
    canvas.drawCircle(w / 2f, h / 2f, radius, backgroundPaint)

    // Draw initials
    val initials = getInitials(name)
    textPaint.textSize = radius * 0.7f
    val textY = h / 2f - (textPaint.descent() + textPaint.ascent()) / 2f
    canvas.drawText(initials, w / 2f, textY, textPaint)
  }

  private fun getInitials(fullName: String): String {
    val parts = fullName.trim().split("\\s+".toRegex())
    return if (parts.size == 1) {
      parts[0].take(2).uppercase()
    } else {
      "${parts.first()[0]}${parts.last()[0]}".uppercase()
    }
  }

  private fun colorFromName(seed: String): Int {
    val hash = seed.fold(0) { acc, ch -> acc + ch.code }
    return avatarColors[hash % avatarColors.size]
  }
}

File: modules/avatar-view/ios/AvatarView.swift
import UIKit
import ExpoModulesCore

public class AvatarSwiftView: ExpoView {

  private var name: String = ""
  private var sizePt: Int = 40

  /// Deterministic color palette — mirrors the JS Avatar component
  private let avatarColors: [UIColor] = [
    UIColor(red: 0.051, green: 0.580, blue: 0.533, alpha: 1), // #0D9488 teal
    UIColor(red: 0.486, green: 0.227, blue: 0.929, alpha: 1), // #7C3AED purple
    UIColor(red: 0.145, green: 0.388, blue: 0.922, alpha: 1), // #2563EB blue
    UIColor(red: 0.937, green: 0.267, blue: 0.267, alpha: 1), // #EF4444 red
    UIColor(red: 0.961, green: 0.620, blue: 0.043, alpha: 1), // #F59E0B amber
    UIColor(red: 0.063, green: 0.725, blue: 0.506, alpha: 1), // #10B981 green
  ]

  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    backgroundColor = .clear
  }

  func setName(_ newName: String) {
    name = newName
    setNeedsDisplay()
  }

  func setSize(_ newSize: Int) {
    sizePt = newSize
    setNeedsDisplay()
  }

  public override func draw(_ rect: CGRect) {
    guard let ctx = UIGraphicsGetCurrentContext() else { return }

    let side = min(rect.width, rect.height)
    let origin = CGPoint(
      x: (rect.width - side) / 2,
      y: (rect.height - side) / 2
    )
    let circleRect = CGRect(origin: origin, size: CGSize(width: side, height: side))

    // Background circle
    ctx.setFillColor(colorFromName(name).cgColor)
    ctx.fillEllipse(in: circleRect)

    // Initials
    let initials = getInitials(name)
    let fontSize = side * 0.35
    let attrs: [NSAttributedString.Key: Any] = [
      .font: UIFont.boldSystemFont(ofSize: fontSize),
      .foregroundColor: UIColor.white,
    ]
    let textSize = (initials as NSString).size(withAttributes: attrs)
    let textRect = CGRect(
      x: circleRect.midX - textSize.width / 2,
      y: circleRect.midY - textSize.height / 2,
      width: textSize.width,
      height: textSize.height
    )
    (initials as NSString).draw(in: textRect, withAttributes: attrs)
  }

  private func getInitials(_ fullName: String) -> String {
    let parts = fullName.trimmingCharacters(in: .whitespaces)
      .components(separatedBy: .whitespaces)
      .filter { !$0.isEmpty }

    if parts.count == 1 {
      return String(parts[0].prefix(2)).uppercased()
    }
    let first = parts.first!.first!
    let last = parts.last!.first!
    return "\(first)\(last)".uppercased()
  }

  private func colorFromName(_ seed: String) -> UIColor {
    let hash = seed.unicodeScalars.reduce(0) { $0 + Int($1.value) }
    return avatarColors[hash % avatarColors.count]
  }
}
